better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1004 @@
+every word a reader has to process has a cognitive cost in addition to
+whatever benefits it may provide, so each one should more than pay for
+itself.
+
+Secondly, the summary has become so simple that we can arguably embed
+the precondition there, making the final declaration:
+
+```swift
+/// Sorts the elements so that all adjacent pairs satisfy the [partial
+/// ordering](https://en.wikipedia.org/wiki/Partially_ordered_set)
+/// `areInOrder`.
+///
+/// - Complexity: at most N log N comparisons, where N is the number
+///   of elements.
+mutating func sort<T>(areInOrder: (T, T)->Bool) { ... }
+```
+
+There is one factor weighing against all these simplifications,
+though: sorting functions taking strict weak orderings are
+traditional. One has to weigh the risk of users reflexively writing
+`sort(areInOrder: <)`—against the value of these simplifications.  The
+argument label helps a bit, because *if* you happen to consider the
+equal elements case, you will see they clearly `areInOrder` but and
+realize `<` will return false.  But if you don't happen to consider
+that case, you won't get the sort you expected… sometimes.  Whether to
+break with precedent in order to get a simpler API with a clearer
+contract is an engineering decision you will have to make. To reduce
+the risk you could add this assertion, which will fire if the ordering
+is strict-weak:
+
+```
+assert(
+  self.isEmpty || areInOrder(first!, first!),
+  "Partial ordering required; did you pass a strict-weak ordering?")
+```
+
+
+### Project-Wide Documentation Policies
+
+The only reason you haven't seen
+complexity documented in this talk up to now is that I have a policy
+that operations have constant complexity unless specifically
+documented otherwise.
+
+Which brings me to this aside…  A big part of making the documentation
+problem tractable is having some well-chosen project-wide policies
+that save you from repeating common patterns. So I'm not going to
+prescribe your project's policy about this, but you should choose one,
+and write it down somewhere.  Your project's choice of policy can make
+the difference between documentation being useful and being burdensome
+or inconsistent (at which point people will just stop reading and
+writing it).
+
+For example,
+
+- Every declaration outside a function body must have a documentation comment that describes
+  its contract.
+  - Start with a summary sentence fragment.
+    - Describe what a function or method does and what it returns.
+    - Describe what a property or type is.
+    - Separate the fragment from any additional documentation with a blank line and end it
+      with a period.
+  - Preconditions, postconditions and invariants obviously implied by the summary need not
+    be explicitly documented.
+  - Declarations that fulfill protocol requirements are exempted when
+    nothing useful can be added to the documentation of the protocol
+    requirement itself.
+
+- End every file with a newline.
```

Comment:
I don't think this is a good example. This kind of thing should be automatically handled by editor and pre-commit tooling so developers don't concern themselves with it.